src/scenes/MerchStorePage/Redemption/useRedeemMerch.ts

```diff
@@ -0,0 +1,118 @@
+import { signMessage } from '@wagmi/core';
+import { useCallback } from 'react';
+import { graphql } from 'react-relay';
+import { SelectorStoreUpdater } from 'relay-runtime';
+import { useAccount } from 'wagmi';
+
+import { useToastActions } from '~/contexts/toast/ToastContext';
+import { useRedeemMerchMutation } from '~/generated/useRedeemMerchMutation.graphql';
+import { usePromisifiedMutation } from '~/hooks/usePromisifiedMutation';
+
+type Props = {
+  tokenIds: string[];
+  onSuccess: () => void;
+};
+
+export default function useRedeemMerch() {
+  const { address } = useAccount();
+
+  const [redeemMerch] = usePromisifiedMutation<useRedeemMerchMutation>(graphql`
+    mutation useRedeemMerchMutation($input: RedeemMerchInput!) @raw_response_type {
+      redeemMerch(input: $input) {
+        __typename
+        ... on RedeemMerchPayload {
+          discountCodes {
+            code
+            tokenId
+          }
+        }
+        ... on ErrInvalidInput {
+          __typename
+          message
+        }
+      }
+    }
+  `);
+
+  const { pushToast } = useToastActions();
+
+  return useCallback(
+    async ({ tokenIds, onSuccess }: Props) => {
+      try {
+        const signature = await signMessage({
+          message: `Gallery uses this cryptographic signature in place of a password: [${tokenIds.join(
+            ', '
+          )}]`,
+        });
+
+        const updater: SelectorStoreUpdater<useRedeemMerchMutation['response']> = (
+          store,
+          response
+        ) => {
+          if (response?.redeemMerch?.__typename === 'RedeemMerchPayload') {
+            const discountCodes = response?.redeemMerch?.discountCodes || [];
+
+            const root = store.get(`client:root:getMerchTokens(wallet:"${address}")`);
+
+            const tokens = root?.getLinkedRecords('tokens') || [];
+
+            // assign the discount code to the token
+            tokens.forEach((token) => {
+              const tokenId = token.getValue('tokenId');
+              const discountCode = discountCodes.find(
+                (discountCode) => discountCode.tokenId === tokenId
+              );
+
+              if (discountCode) {
+                token.setValue(discountCode.code, 'discountCode').setValue(true, 'redeemed');
+              }
+            });
+          }
+        };
```

Comment:
We wouldn't have to write any of this code if the mutation payload looked like this
```graphql
type RedeemMerchPayload {
    # The client can select `discountCode` off of this and the cache will automatically update
    tokens: [Token]
}
```

Reply:
CC @benny-conn can you implement this change before we land this PR